servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/PayloadSizeLimitingHttpRequesterFilter.java

```diff
@@ -0,0 +1,109 @@
+
+/**
+ * Limits the response payload size for a request. The filter will throw an exception which may result in
+ * stream/connection closure.
+ */
+public final class PayloadSizeLimitingHttpRequesterFilter implements
+                        StreamingHttpClientFilterFactory, StreamingHttpConnectionFilterFactory,
+                        HttpExecutionStrategyInfluencer {
+    private final int maxResponsePayloadSize;
+
+    /**
+     * Create a new instance.
+     * @param maxResponsePayloadSize The maximum response payload size allowed.
+     */
+    public PayloadSizeLimitingHttpRequesterFilter(int maxResponsePayloadSize) {
+        if (maxResponsePayloadSize < 0) {
+            throw new IllegalArgumentException("maxResponsePayloadSize: " + maxResponsePayloadSize + " (expected >0)");
+        }
+        this.maxResponsePayloadSize = maxResponsePayloadSize;
+    }
+
+    @Override
+    public StreamingHttpClientFilter create(final FilterableStreamingHttpClient client) {
+        return new StreamingHttpClientFilter(client) {
+            @Override
+            protected Single<StreamingHttpResponse> request(final StreamingHttpRequester delegate,
+                                                            final StreamingHttpRequest request) {
+                return applyLimit(request, delegate::request);
+            }
+        };
+    }
+
+    @Override
+    public StreamingHttpConnectionFilter create(final FilterableStreamingHttpConnection connection) {
+        return new StreamingHttpConnectionFilter(connection) {
+            @Override
+            public Single<StreamingHttpResponse> request(final StreamingHttpRequest request) {
+                return applyLimit(request, super::request);
+            }
+        };
+    }
+
+    @Override
+    public HttpExecutionStrategy requiredOffloads() {
+        return offloadNone();
+    }
+
+    private Single<StreamingHttpResponse> applyLimit(
+            StreamingHttpRequest request, Function<StreamingHttpRequest, Single<StreamingHttpResponse>> delegator) {
+        return delegator.apply(request).map(response -> response.transformPayloadBody(publisher ->
+                Publisher.defer(() -> {
+                    final MutableInt responsePayloadSize = new MutableInt();
+                    return publisher.beforeOnNext(buff -> {
+                        if (maxResponsePayloadSize - responsePayloadSize.value < buff.readableBytes()) {
+                            throw new PayloadTooLongException("Maximum payload size=" + maxResponsePayloadSize +
+                                    " current payload size=" + responsePayloadSize.value +
+                                    " new buffer size=" + buff.readableBytes());
+                        }
+                        responsePayloadSize.value += buff.readableBytes();
+                    }).shareContextOnSubscribe();
+                })));
+    }
+
+    static final class PayloadTooLongException extends IllegalStateException {
```

Comment:
do we need the stack-trace here?

Reply:
From javadoc for `IllegalStateException`:
> Signals that a method has been invoked at an illegal or inappropriate time. In other words, the Java environment or Java application is not in an appropriate state for the requested operation.

Hard to say that `PayloadTooLongException` is about invoking at illegal or inappropriate time. Have you considered `SerializationException` or `IllegalArgumentException` instead?

Reply:
I will extend from `RuntimeException` for now